### PR TITLE
[SYCL] Add more cleanup points close to the shutdown

### PR DESCRIPTION
**Pull request by KseniyaTikhomirova**
Previously we tried to release internal resources only when the last (likely main) thread exits. For multi threaded applications it could be not functional. e.g. threads are closed during module unload and when main thread exits - other threads are still alive. Adding logic to try non blocking release for every thread exit and blocking for the last one.

### CODE REVIEW

File: sycl/source/detail/global_handler.cpp [hunk 1]
```diff
@@ -53,12 +53,10 @@ class ObjectUsageCounter {
...
-    if (!MCounter) {
-      LockGuard Guard(GlobalHandler::MSyclGlobalHandlerProtector);
-      GlobalHandler *RTGlobalObjHandler = GlobalHandler::getInstancePtr();
-      if (RTGlobalObjHandler) {
-        RTGlobalObjHandler->prepareSchedulerToRelease();
-      }
+    LockGuard Guard(GlobalHandler::MSyclGlobalHandlerProtector);
+    GlobalHandler *RTGlobalObjHandler = GlobalHandler::getInstancePtr();
+    if (RTGlobalObjHandler) {
+      RTGlobalObjHandler->prepareSchedulerToRelease(!MCounter);
```
Comment (againull) [inline sycl/source/detail/global_handler.cpp:59]: Should we put `MCounter-- ` under the Guard? If MCounter is decremented by two threads approximately at the same time then we will have MCounter == 0 for two threads, so we will call RTGlobalObjHandler->prepareSchedulerToRelease(true) twice. Is this a mistake or is it safe to call prepareSchedulerToRelease(true) twice? Probably I am missing something.

Comment (KseniyaTikhomirova) [reply sycl/source/detail/global_handler.cpp:59]: it should not be a problem but I would prefer to fix this. I haven't noticed that. Thank you!

Comment (KseniyaTikhomirova) [reply sycl/source/detail/global_handler.cpp]: fixed here https://github.com/intel/llvm/pull/11119/commits/52a15f46cf33b16b917b6cc3c3a69b21ae235006

**Review by againull — APPROVED** [review]

**Comment by KseniyaTikhomirova** [comment]
fix for wrong loop condition is here https://github.com/intel/llvm/pull/11791

**Pull request merged**